src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+/*
+ * Copyright (c) 2016 - present Adobe Systems Incorporated. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define, WeakMap, Uint32Array, Uint8Array */
+
+define(function (require, exports, module) {
+    "use strict";
+    var StringUtils = require("utils/StringUtils"),
+        _           = require("thirdparty/lodash");
+
+    /**
+     * Store document text and index
+     * key: {CodeMirror.Doc}
+     * value: { text: {string}, index: {Array<number>}, generation: {number} }
+     * text = The text of the document
+     * index = The document line index for lookups
+     * generation = the document history generation number
+     */
+    var _documentMap = new WeakMap();
+
+    /**
+     * Determines if the current document has changed since we last stored the docInfo
+     * @param {CodeMirror.Doc} doc
+     * @return boolean
+     */
+    function _needToIndexDocument(doc) {
+        var docInfo = _documentMap.get(doc);
+
+        // lastmodtime is not changed when undo is invoked.
+        // so we will use the generation count to determine if the document has changed
+        if (docInfo && (docInfo.generation === doc.history.generation)) {
+            // document has not changed since we indexed
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Creates an array which stores the sum of all characters in the document
+     * up to the point of each line.
+     * This is needed to efficiently convert character index offsets to position objects of line and character offset.
+     * @param {String} text The string to index
+     * @param {String} lineSeparator The ending character that splits lines
+     */
+    function _createLineCharacterCountIndex(text, lineSeparator) {
```

Comment:
Referring to the same index by different names (`index`, `lineCharacterCountIndex`, `getDocumentIndex`) is a little bit confusing. What the index in question really is, it is the offset, if characters, of the line number `n` in the document. So two points: 1. We need to use a single name consistently; 2. I suggest it could be `lineOffsets`.